### ISSUE TITLE
[FEATURE] Avoid exec as liveness probe

### ISSUE THREAD

**Issue by ElisaMeng** [longhorn/longhorn#2294]
**Is your feature request related to a problem? Please describe.**

According to this https://github.com/opencontainers/runc/issues/2532 
liveness probe with exec for every seconds seem consume too much resource. 

Can we avoid liveness probe with exec in instance managers?

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by shuo-wu** [comment]
Actually we had encountered this issue then modified the period to 5s from 1s: https://github.com/longhorn/longhorn/issues/767

Maybe we still need further investigation for this liveness probe.

**Comment by ElisaMeng** [comment]
@shuo-wu I suspect it leads the system into #2297

**Comment by ElisaMeng** [comment]
> Actually we had encountered this issue then modified the period to 5s from 1s: #767
> 
> Maybe we still need further investigation for this liveness probe.

grpc-health-probe default timeout is 1s, and longhorn probe every seconds. In case server is not reachable, longhorn actually make the second probe before the first one completed. I don't if this will mess up k3s and runc runtime in some way.

**Comment by shuo-wu** [comment]
> > Actually we had encountered this issue then modified the period to 5s from 1s: #767
> > Maybe we still need further investigation for this liveness probe.
> 
> grpc-health-probe default timeout is 1s, and longhorn probe every seconds. In case server is not reachable, longhorn actually make the second probe before the first one completed. I don't if this will mess up k3s and runc runtime in some way.

For the gRPC health probe, there are 2 timeouts:
```
Usage of grpc_health_probe:
  -addr string
    	(required) tcp host:port to connect
  -connect-timeout duration
    	timeout for establishing connection (default 1s)
  -rpc-timeout duration
    	timeout for health check rpc (default 1s)
```
Based on this info, we can set the Longhorn liveness probe `TimeoutSeconds` to 2s, then there won't be more than 1 probe running. But I don't think it is the root cause since thee liveness probe could start 2 probes at max with the current setting.

**Comment by ElisaMeng** [comment]
> 
> Based on this info, we can set the Longhorn liveness probe `TimeoutSeconds` to 2s, then there won't be more than 1 probe running. But I don't think it is the root cause since thee liveness probe could start 2 probes at max with the current setting.

Providing runc as it should, I agree with you, but if runc takes a bit long time to clean up exec, would it be the case?

**Comment by shuo-wu** [comment]
I am not sure... It's better to know what these `runc`s were doing.

**Comment by cclhsu** [comment]
@ElisaMeng It seems like the feature is related to https://github.com/longhorn/longhorn/issues/2297?

Currently, We have probe configurations as below

```
initialdelayseconds: 3
timeoutseconds: 1
periodseconds: 5
successthreshold: 1
failurethreshold: 3
```

The configuration covered the slow starting container ` set up a startup probe with the same command with a failureThreshold * periodSeconds long enough to cover the worse case startup time` as described in [Configure Liveness, Readiness and Startup Probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).  However, the same document has notes and caution as below: 

```
Note:
Before Kubernetes 1.20, the field timeoutSeconds was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.

This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior, even without realizing it, as the default timeout is 1 second. As a cluster administrator, you can disable the feature gate ExecProbeTimeout (set it to false) on each kubelet to restore the behavior from older versions, then remove that override once all the exec probes in the cluster have a timeoutSeconds value set.
If you have pods that are impacted from the default 1 second timeout, you should update their probe timeout so that you're ready for the eventual removal of that feature gate.

With the fix of the defect, for exec probes, on Kubernetes 1.20+ with the dockershim container runtime, the process inside the container may keep running even after probe returned failure because of the timeout.
```

```
Caution: Incorrect implementation of readiness probes may result in an ever growing number of processes in the container, and resource starvation if this is left unchecked.
```

We will continue to monitor the behavior of liveness probe as we update Kubernetes runtime to v1.20+.  However, maybe we can close the issue for now since the root cause is caused by the mal-formulate script?

**Comment by jenting** [comment]
> @ElisaMeng It seems like the feature is related to #2297?
> 
> Currently, We have probe configurations as below
> 
> ```
> initialdelayseconds: 3
> timeoutseconds: 1
> periodseconds: 5
> successthreshold: 1
> failurethreshold: 3
> ```
> 
> The configuration covered the slow starting container ` set up a startup probe with the same command with a failureThreshold * periodSeconds long enough to cover the worse case startup time` as described in [Configure Liveness, Readiness and Startup Probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/). However, the same document has notes and caution as below:
> 
> ```
> Note:
> Before Kubernetes 1.20, the field timeoutSeconds was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.
> 
> This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior, even without realizing it, as the default timeout is 1 second. As a cluster administrator, you can disable the feature gate ExecProbeTimeout (set it to false) on each kubelet to restore the behavior from older versions, then remove that override once all the exec probes in the cluster have a timeoutSeconds value set.
> If you have pods that are impacted from the default 1 second timeout, you should update their probe timeout so that you're ready for the eventual removal of that feature gate.
> 
> With the fix of the defect, for exec probes, on Kubernetes 1.20+ with the dockershim container runtime, the process inside the container may keep running even after probe returned failure because of the timeout.
> ```
> 
> ```
> Caution: Incorrect implementation of readiness probes may result in an ever growing number of processes in the container, and resource starvation if this is left unchecked.
> ```
> 
> We will continue to monitor the behavior of liveness probe as we update Kubernetes runtime to v1.20+. However, maybe we can close the issue for now since the root cause is caused by the mal-formulate script?

This issue is tackled by #2699. So we consider it as fixed. Close it.
